Gathering data

In [1]:
import pandas as pd
import numpy as np
import requests
import os
import tweepy as tw
import json

In [2]:
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')

In [3]:
twitter_archive.head(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None  
2                  10   Archie  None    None   None  None

In [4]:
folder_name = 'image_prediction'
if not os.path.exists(folder_name):
    os.makedirs(folder_name)

In [5]:
#programmatically download the tsv file
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
response

<Response [200]>

In [6]:
#save to a folder
with open(os.path.join(folder_name, url.split('/')[-1]), mode='wb') as file:
    file.write(response.content)

In [7]:
image_prediction = pd.read_csv('image_prediction/image-predictions.tsv', delimiter= '\t')
image_prediction

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0           1  Welsh_springer_spaniel  0.465074    True              collie   
1           1                 redbone  0.506826    True  miniature_pinscher   
2           1         German_shepherd  0.596461    True            malinois   
3           1     Rhodesian_ridgeback  0.408143    True             redbone   
4           1      miniature_pinscher  0.560311    True          Rottweiler   
...       ...                     ...       ...     ...                 ...   
2070        2                  basset  0.555712    True    English_springer   
2071        1             paper_towel  0.170278   False  Labrador_retriever   
2072        1               Chihuahua  0.716012    True            malamute   
2073        1               Chihuahua  0.323581    True            Pekinese   
2074        1                  orange  0.097049   False               bagel   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665    True            Shetland_sheepdog  0.061428    True  
1     0.074192    True          Rhodesian_ridgeback  0.072010    True  
2     0.138584    True                   bloodhound  0.116197    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
4     0.243682    True                     Doberman  0.154629    True  
...        ...     ...                          ...       ...     ...  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2071  0.168086    True                      spatula  0.040836   False  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  
2074  0.085851   False                       banana  0.076110   False  

[2075 rows x 12 columns]

In [8]:
consumer_key= 'BIi26LduwxyflyqYSTrb8aLDN'
consumer_secret= 'bsPax1bDvD8kMjObVVLG8esh6jFtyURBvyFuffRgKu0lmA0jYk'
access_token= '1483108734406696962-Nl9iW4Xptb2BdVa2brtOV3FbalbTo5'
access_token_secret= 'ICTF3hLUi7eeN64tkfilUpMTdy4lBZHhGdN3r9QpC0Cb0'
bearer_token = 'AAAAAAAAAAAAAAAAAAAAAKMKeQEAAAAAda13MQ8VYBJQP8PwkZeQQDd40ag%3DNaczZkQEDrR6s4x0bhARjIGkXiXCXvLFB439xX6w0EJ6Paf65A'

In [9]:
auth = tw.OAuth1UserHandler(consumer_key, consumer_secret, access_token, access_token_secret)
api = tw.API(auth,wait_on_rate_limit= True)
#use the JSON parser to generate json in the form of dictionaries from the API
#parser=(),tw.parsers.JSONParser,

In [10]:
tweet_ids = list(twitter_archive.tweet_id)

In [11]:
tweet_ids

[892420643555336193,
 892177421306343426,
 891815181378084864,
 891689557279858688,
 891327558926688256,
 891087950875897856,
 890971913173991426,
 890729181411237888,
 890609185150312448,
 890240255349198849,
 890006608113172480,
 889880896479866881,
 889665388333682689,
 889638837579907072,
 889531135344209921,
 889278841981685760,
 888917238123831296,
 888804989199671297,
 888554962724278272,
 888202515573088257,
 888078434458587136,
 887705289381826560,
 887517139158093824,
 887473957103951883,
 887343217045368832,
 887101392804085760,
 886983233522544640,
 886736880519319552,
 886680336477933568,
 886366144734445568,
 886267009285017600,
 886258384151887873,
 886054160059072513,
 885984800019947520,
 885528943205470208,
 885518971528720385,
 885311592912609280,
 885167619883638784,
 884925521741709313,
 884876753390489601,
 884562892145688576,
 884441805382717440,
 884247878851493888,
 884162670584377345,
 883838122936631299,
 883482846933004288,
 883360690899218434,
 883117836046

In [12]:
status1 = api.get_status(tweet_ids[0])
status1.id

892420643555336193

In [13]:
data = []
for tweet in tweet_ids:
    try:
        status = api.get_status(tweet)
        data.append([status.id,status.favorite_count, status.retweet_count])
    except:
        pass

Rate limit reached. Sleeping for: 366
Rate limit reached. Sleeping for: 351


In [14]:
status

Status(_api=<tweepy.api.API object at 0x0000028232322280>, _json={'created_at': 'Sun Nov 15 22:32:08 +0000 2015', 'id': 666020888022790149, 'id_str': '666020888022790149', 'text': 'Here we have a Japanese Irish Setter. Lost eye in Vietnam (?). Big fan of relaxing on stair. 8/10 would pet https://t.co/BLDqew2Ijj', 'truncated': False, 'entities': {'hashtags': [], 'symbols': [], 'user_mentions': [], 'urls': [], 'media': [{'id': 666020881337073664, 'id_str': '666020881337073664', 'indices': [108, 131], 'media_url': 'http://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg', 'media_url_https': 'https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg', 'url': 'https://t.co/BLDqew2Ijj', 'display_url': 'pic.twitter.com/BLDqew2Ijj', 'expanded_url': 'https://twitter.com/dog_rates/status/666020888022790149/photo/1', 'type': 'photo', 'sizes': {'medium': {'w': 960, 'h': 720, 'resize': 'fit'}, 'thumb': {'w': 150, 'h': 150, 'resize': 'crop'}, 'large': {'w': 960, 'h': 720, 'resize': 'fit'}, 'small': {'w': 680, 'h': 510

In [15]:
data

[[892420643555336193, 33777, 6995],
 [892177421306343426, 29298, 5295],
 [891815181378084864, 22031, 3475],
 [891689557279858688, 36893, 7217],
 [891327558926688256, 35276, 7747],
 [891087950875897856, 17789, 2596],
 [890971913173991426, 10352, 1658],
 [890729181411237888, 56803, 15737],
 [890609185150312448, 24492, 3615],
 [890240255349198849, 27920, 6090],
 [890006608113172480, 27007, 6143],
 [889880896479866881, 24541, 4159],
 [889665388333682689, 42001, 8338],
 [889638837579907072, 23642, 3709],
 [889531135344209921, 13347, 1878],
 [889278841981685760, 22093, 4439],
 [888917238123831296, 25609, 3757],
 [888804989199671297, 22430, 3529],
 [888554962724278272, 17302, 2876],
 [888078434458587136, 19136, 2892],
 [887705289381826560, 26594, 4531],
 [887517139158093824, 40678, 9838],
 [887473957103951883, 60134, 14999],
 [887343217045368832, 29567, 8805],
 [887101392804085760, 26950, 4982],
 [886983233522544640, 30346, 6303],
 [886736880519319552, 10489, 2625],
 [886680336477933568, 1971

In [16]:
columns = ['tweet_id', 'favorite_count', 'retweet_count']
tweets = pd.DataFrame(data, columns= columns)
tweets

tweet_id  favorite_count  retweet_count
0     892420643555336193           33777           6995
1     892177421306343426           29298           5295
2     891815181378084864           22031           3475
3     891689557279858688           36893           7217
4     891327558926688256           35276           7747
...                  ...             ...            ...
2322  666049248165822465              88             36
2323  666044226329800704             247            115
2324  666033412701032449             100             36
2325  666029285002620928             112             39
2326  666020888022790149            2290            422

[2327 rows x 3 columns]

In [17]:
tweets.to_csv('tweets.csv')
twitter_archive.to_csv('twitter_archive.csv')
image_prediction.to_csv('image_prediction.csv')
#saved the dataframe for personal purpose.

## Assessment

In [18]:
twitter_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

In [19]:
image_prediction

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0           1  Welsh_springer_spaniel  0.465074    True              collie   
1           1                 redbone  0.506826    True  miniature_pinscher   
2           1         German_shepherd  0.596461    True            malinois   
3           1     Rhodesian_ridgeback  0.408143    True             redbone   
4           1      miniature_pinscher  0.560311    True          Rottweiler   
...       ...                     ...       ...     ...                 ...   
2070        2                  basset  0.555712    True    English_springer   
2071        1             paper_towel  0.170278   False  Labrador_retriever   
2072        1               Chihuahua  0.716012    True            malamute   
2073        1               Chihuahua  0.323581    True            Pekinese   
2074        1                  orange  0.097049   False               bagel   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665    True            Shetland_sheepdog  0.061428    True  
1     0.074192    True          Rhodesian_ridgeback  0.072010    True  
2     0.138584    True                   bloodhound  0.116197    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
4     0.243682    True                     Doberman  0.154629    True  
...        ...     ...                          ...       ...     ...  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2071  0.168086    True                      spatula  0.040836   False  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  
2074  0.085851   False                       banana  0.076110   False  

[2075 rows x 12 columns]

In [20]:
tweets

tweet_id  favorite_count  retweet_count
0     892420643555336193           33777           6995
1     892177421306343426           29298           5295
2     891815181378084864           22031           3475
3     891689557279858688           36893           7217
4     891327558926688256           35276           7747
...                  ...             ...            ...
2322  666049248165822465              88             36
2323  666044226329800704             247            115
2324  666033412701032449             100             36
2325  666029285002620928             112             39
2326  666020888022790149            2290            422

[2327 rows x 3 columns]

In [21]:
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [22]:
twitter_archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [23]:
#the describe function shows the maximum denominator to be greater than 10. This is a quality data issue.
twitter_archive.query('rating_denominator > 10')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
342   832088576586297345           8.320875e+17         3.058208e+07   
433   820690176645140481                    NaN                  NaN   
784   775096608509886464                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
1068  740373189193256964                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1165  722974582966214656                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1635  684222868335505415                    NaN                  NaN   
1662  682962037429899265                    NaN                  NaN   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1779  677716515794329600                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   

                      timestamp  \
342   2017-02-16 04:45:50 +0000   
433   2017-01-15 17:52:40 +0000   
784   2016-09-11 22:20:06 +0000   
902   2016-07-28 01:00:57 +0000   
1068  2016-06-08 02:41:38 +0000   
1120  2016-05-13 16:15:54 +0000   
1165  2016-04-21 02:25:47 +0000   
1202  2016-04-03 01:36:11 +0000   
1228  2016-03-27 01:29:02 +0000   
1254  2016-03-18 02:46:49 +0000   
1274  2016-03-14 02:04:08 +0000   
1351  2016-02-28 21:25:30 +0000   
1433  2016-02-10 16:51:59 +0000   
1598  2016-01-10 04:04:10 +0000   
1634  2016-01-05 04:11:44 +0000   
1635  2016-01-05 04:00:18 +0000   
1662  2016-01-01 16:30:13 +0000   
1663  2016-01-01 06:22:03 +0000   
1779  2015-12-18 05:06:23 +0000   
1843  2015-12-13 01:41:41 +0000   

                                                 source  \
342   <a href="http://twitter.com/download/iphone" r...   
433   <a href="http://twitter.com/download/iphone" r...   
784   <a href="http://twitter.com/download/iphone" r...   
902   <a href="http://twitter.com/download/iphone" r...   
1068  <a href="http://twitter.com/download/iphone" r...   
1120  <a href="http://twitter.com/download/iphone" r...   
1165  <a href="http://twitter.com/download/iphone" r...   
1202  <a href="http://twitter.com/download/iphone" r...   
1228  <a href="http://twitter.com/download/iphone" r...   
1254  <a href="http://twitter.com/download/iphone" r...   
1274  <a href="http://twitter.com/download/iphone" r...   
1351  <a href="http://twitter.com/download/iphone" r...   
1433  <a href="http://twitter.com/download/iphone" r...   
1598  <a href="http://twitter.com/download/iphone" r...   
1634  <a href="http://twitter.com/download/iphone" r...   
1635  <a href="http://twitter.com/download/iphone" r...   
1662  <a href="http://twitter.com/download/iphone" r...   
1663  <a href="http://twitter.com/download/iphone" r...   
1779  <a href="http://twitter.com/download/iphone" r...   
1843  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
342            @docmisterio account started on 11/15/15                  NaN   
433   The floofs have been released I repeat the flo...                  NaN   
784   RT @dog_rates: After so many requests, this is...         7.403732e+17   
902   Why does this never happen at my front door......                  NaN   
1068  After so many requests, this is Bretagne. She ...                  NaN   
1120  Say hello to this unbelievably well behav

In [24]:
#The describe function shows the maximum numerator to be 1776 which seems to be unusual.
twitter_archive.query('rating_numerator == 1776')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
979  749981277374128128                    NaN                  NaN   

                     timestamp  \
979  2016-07-04 15:00:45 +0000   

                                                source  \
979  <a href="https://about.twitter.com/products/tw...   

                                                  text  retweeted_status_id  \
979  This is Atticus. He's quite simply America af....                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
979                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
979  https://twitter.com/dog_rates/status/749981277...              1776   

     rating_denominator     name doggo floofer pupper puppo  
979                  10  Atticus  None    None   None  None

In [25]:
twitter_archive.sample(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
743   780459368902959104                    NaN                  NaN   
719   783391753726550016                    NaN                  NaN   
2197  668826086256599040                    NaN                  NaN   

                      timestamp  \
743   2016-09-26 17:29:48 +0000   
719   2016-10-04 19:42:03 +0000   
2197  2015-11-23 16:18:59 +0000   

                                                 source  \
743   <a href="http://twitter.com/download/iphone" r...   
719   <a href="http://twitter.com/download/iphone" r...   
2197  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
743   This is Bear. Don't worry, he's not a real bea...                  NaN   
719   Meet Godi. He's an avid beachgoer and part tim...                  NaN   
2197  This is Banditt. He is a brown LaBeouf retriev...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
743                        NaN                        NaN   
719                        NaN                        NaN   
2197                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
743   https://twitter.com/dog_rates/status/780459368...                11   
719   https://twitter.com/dog_rates/status/783391753...                13   
2197  https://twitter.com/dog_rates/status/668826086...                10   

      rating_denominator     name doggo floofer pupper puppo  
743                   10     Bear  None    None   None  None  
719                   10     Godi  None    None   None  None  
2197                  10  Banditt  None    None   None  None

In [26]:
twitter_archive.expanded_urls.sample(3)

1200    https://twitter.com/chpsanfrancisco/status/716...
1102    https://twitter.com/dog_rates/status/735274964...
492     https://twitter.com/dog_rates/status/813217897...
Name: expanded_urls, dtype: object

In [27]:
twitter_archive.isnull().sum()
#missing values

tweet_id                         0
in_reply_to_status_id         2278
in_reply_to_user_id           2278
timestamp                        0
source                           0
text                             0
retweeted_status_id           2175
retweeted_status_user_id      2175
retweeted_status_timestamp    2175
expanded_urls                   59
rating_numerator                 0
rating_denominator               0
name                             0
doggo                            0
floofer                          0
pupper                           0
puppo                            0
dtype: int64

In [28]:
image_prediction.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [29]:
image_prediction.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [30]:
image_prediction.sample(3)

tweet_id                                            jpg_url  \
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
166   668981893510119424    https://pbs.twimg.com/media/CUize-0WEAAerAK.jpg   
1471  779377524342161408  https://pbs.twimg.com/ext_tw_video_thumb/77937...   

      img_num                   p1   p1_conf  p1_dog            p2   p2_conf  \
3           1  Rhodesian_ridgeback  0.408143    True       redbone  0.360687   
166         1            jellyfish  0.447246   False    coral_reef  0.238625   
1471        1              sundial  0.170921   False  cash_machine  0.060359   

      p2_dog                  p3   p3_conf  p3_dog  
3       True  miniature_pinscher  0.222752    True  
166    False            goldfish  0.040227   False  
1471   False                maze  0.054981   False

In [31]:
image_prediction.query('p1_dog == False')
#p1_dog is the most accurate prediction for the images.

tweet_id                                            jpg_url  \
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
...                  ...                                                ...   
2026  882045870035918850    https://pbs.twimg.com/media/DD2oCl2WAAEI_4a.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2071  891689557279858688    https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2074  892420643555336193    https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                p1   p1_conf  p1_dog                  p2  \
6           1        box_turtle  0.933012   False          mud_turtle   
8           1     shopping_cart  0.962465   False     shopping_basket   
17          1               hen  0.965932   False                cock   
18          1  desktop_computer  0.086502   False                desk   
21          1  three-toed_sloth  0.914671   False               otter   
...       ...               ...       ...     ...                 ...   
2026        1          web_site  0.949591   False               dhole   
2046        1       convertible  0.738995   False          sports_car   
2052        1         limousine  0.130432   False           tow_truck   
2071        1       paper_towel  0.170278   False  Labrador_retriever   
2074        1            orange  0.097049   False               bagel   

       p2_conf  p2_dog                p3   p3_conf  p3_dog  
6     0.045885   False          terrapin  0.017885   False  
8     0.014594   False  golden_retriever  0.007959    True  
17    0.033919   False         partridge  0.000052   False  
18    0.085547   False          bookcase  0.079480   False  
21    0.015250   False    great_grey_owl  0.013207   False  
...        ...     ...               ...       ...     ...  
2026  0.017326   False  golden_retriever  0.006941    True  
2046  0.139952   False         car_wheel  0.044173   False  
2052  0.029175   False     shopping_cart  0.026321   False  
2071  0.168086    True           spatula  0.040836   False  
2074  0.085851   False            banana  0.076110   False  

[543 rows x 12 columns]

In [32]:
image_prediction.isnull().sum()
#no missing values

tweet_id    0
jpg_url     0
img_num     0
p1          0
p1_conf     0
p1_dog      0
p2          0
p2_conf     0
p2_dog      0
p3          0
p3_conf     0
p3_dog      0
dtype: int64

In [33]:
tweets.sample(5)

tweet_id  favorite_count  retweet_count
1507  689977555533848577            1271            410
1914  673709992831262724             752            236
248   842765311967449089            6300           1176
1390  698355670425473025            1703            396
792   771014301343748096            6482           1536

In [34]:
tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2327 entries, 0 to 2326
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   tweet_id        2327 non-null   int64
 1   favorite_count  2327 non-null   int64
 2   retweet_count   2327 non-null   int64
dtypes: int64(3)
memory usage: 54.7 KB


In [35]:
tweets.isnull().sum()
#no missing values

tweet_id          0
favorite_count    0
retweet_count     0
dtype: int64

In [36]:
tweets.describe()

tweet_id  favorite_count  retweet_count
count  2.327000e+03     2327.000000    2327.000000
mean   7.417930e+17     7046.259991    2467.479158
std    6.820795e+16    10949.877347    4178.579011
min    6.660209e+17        0.000000       1.000000
25%    6.781394e+17     1224.000000     493.500000
50%    7.178418e+17     3047.000000    1148.000000
75%    7.986547e+17     8593.000000    2856.000000
max    8.924206e+17   144708.000000   70617.000000

### Quality Issues

#### twitter_archive
    1.'in_reply_to_status','in_reply_to_user_id','retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp' all have missing data.
    2. Missing data in 'expanded_urls' column.
    3. Missing items in 'in_reply_to_user_id' column.
    4. Data type for 'timestamp' column should be date_time.
    5. 'source' column data are mostly duplicated.
    6. Redundant/ not required columns - 'retweet_status_id', 'retweet_status_time_stamp'.
    7. 'Rating_denominator' have values greater other than 10.
    8. Very high rating 'rating_numerator of 1776 @ row 979.
    9. Dog_name consists of missing data and other names such as 'a'. 
    10. Change data type for retweet and favorite count after merging from float to int.
#### image_predictions

    11. column_names not explanatory.
    12. Names of dog breed predicted separated by under_scores.
    13. 543 rows contains False dog predictions.

### Tidiness Issues
#### twitter_archive
1. 'doggo', 'floofer', 'pupper', 'puppo' should be collapse into a single column - 'dog_stage' (every variable should form a column.)

#### tweets
2. tweets dataframe should be a part of the twitter_archive dataframe. Every observational unit should form a table. The two dataframes have same/similar observation units.

## Cleaning

In [37]:
twitter_archive_clean = twitter_archive.copy()
image_prediction_clean = image_prediction.copy()
tweets_clean = tweets.copy()

### Define, Code, Test
Do this for each assessed issues.

### Tidiness 

tweets dataframe should be a part of the twitter_archive dataframe. Every observational unit should form a table. The two dataframes have same/similar observation units. Use pandas merge function to join two dataframes.

In [420]:
twitter_archive_clean.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None

In [421]:
twitter_archive_clean = pd.merge(twitter_archive_clean, tweets_clean, on= ['tweet_id'], how= 'left')
twitter_archive_clean.sample(10)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1962  673355879178194945                    NaN                  NaN   
1845  675845657354215424                    NaN                  NaN   
373   828376505180889089                    NaN                  NaN   
563   802572683846291456                    NaN                  NaN   
1338  705066031337840642                    NaN                  NaN   
1279  708845821941387268                    NaN                  NaN   
1023  746521445350707200                    NaN                  NaN   
632   793962221541933056                    NaN                  NaN   
1787  677565715327688705                    NaN                  NaN   
47    883117836046086144                    NaN                  NaN   

                      timestamp  \
1962  2015-12-06 04:18:46 +0000   
1845  2015-12-13 01:12:15 +0000   
373   2017-02-05 22:55:23 +0000   
563   2016-11-26 18:00:13 +0000   
1338  2016-03-02 16:23:36 +0000   
1279  2016-03-13 02:43:08 +0000   
1023  2016-06-25 01:52:36 +0000   
632   2016-11-02 23:45:19 +0000   
1787  2015-12-17 19:07:09 +0000   
47    2017-07-07 00:17:54 +0000   

                                                 source  \
1962  <a href="http://twitter.com/download/iphone" r...   
1845  <a href="http://twitter.com/download/iphone" r...   
373   <a href="http://twitter.com/download/iphone" r...   
563   <a href="http://twitter.com/download/iphone" r...   
1338  <a href="http://twitter.com/download/iphone" r...   
1279  <a href="http://twitter.com/download/iphone" r...   
1023  <a href="http://twitter.com/download/iphone" r...   
632   <a href="http://twitter.com/download/iphone" r...   
1787  <a href="http://twitter.com/download/iphone" r...   
47    <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1962  This is Koda. She's a boss. Helps shift gears....                  NaN   
1845  This is Pepper. She's not fully comfortable ri...                  NaN   
373   This is Beebop. Her name means "Good Dog" in r...                  NaN   
563   This is Winnie. She's h*ckin ferocious. Dandel...                  NaN   
1338  This is Reese. He's a Chilean Sohcahtoa. Loves...                  NaN   
1279  Here's a pupper with magic eyes. Not wearing a...                  NaN   
1023  RT @dog_rates: This is Shaggy. He knows exactl...         6.678667e+17   
632   This is Maximus. His face is stuck like that. ...                  NaN   
1787  Contortionist pup here. Inside pentagram. Clea...                  NaN   
47    Please only send dogs. We don't rate mechanics...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1962                       NaN                        NaN   
1845                       NaN                        NaN   
373                        NaN                        NaN   
563                        NaN                        NaN   
1338                       NaN                        NaN   
1279                       NaN                        NaN   
1023              4.196984e+09  2015-11-21 00:46:50 +0000   
632                        NaN                        NaN   
1787                       NaN                        NaN   
47                         NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1962  https://twitter.com/dog_rates/status/673355879...                11   
1845  https://twitter.com/dog_rates/status/675845657...                10   
373   https://twitter.com/dog_rates/status/828376505...                13   
563   https://twitter.com/dog_rates/status/802572683...                12   
1338  https://twitter.com/dog_rates/status/705066031...                12   
1279  https://twitter.com/dog_rates/status/708845821...                 9   
1023  https://twitter.com/dog_rates/status/667866724...                10   
632   https://twit

In [422]:
twitter_archive_clean.shape

(2356, 19)

In [423]:
twitter_archive_clean.retweet_count.isnull().sum()

31

In [424]:
twitter_archive_clean['retweet_count'].dropna()
twitter_archive_clean.retweet_count.isnull().sum()

31

In [425]:
twitter_archive_clean.query('retweet_count == 0')

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo, favorite_count, retweet_count]
Index: []

'doggo', 'floofer', 'pupper', 'puppo' should be collapse into a single column - 'dog_stage' (every variable should form a column.)
Use melt function to collapse the columns to one, and drop the resluting unnecessary columns.

In [426]:
twitter_archive_clean.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator',
       'rating_denominator', 'name', 'doggo', 'floofer', 'pupper', 'puppo',
       'favorite_count', 'retweet_count'],
      dtype='object')

In [427]:
columns_needed = ['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator',
       'rating_denominator','name','favorite_count','retweet_count']
twitter_archive_clean= pd.melt(twitter_archive_clean, id_vars= columns_needed, 
                               value_vars=['doggo', 'floofer', 'pupper', 'puppo'],var_name= 'dog_stage')

In [428]:
twitter_archive_clean.sample()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
5871  724046343203856385                    NaN                  NaN   

                      timestamp  \
5871  2016-04-24 01:24:35 +0000   

                                                 source  \
5871  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
5871  This is Sarge. Not even he knows what his tong...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
5871                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
5871  https://twitter.com/dog_rates/status/724046343...                10   

      rating_denominator   name  favorite_count  retweet_count dog_stage value  
5871                  10  Sarge          2453.0          513.0    pupper  None

In [429]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9424 entries, 0 to 9423
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    9424 non-null   int64  
 1   in_reply_to_status_id       312 non-null    float64
 2   in_reply_to_user_id         312 non-null    float64
 3   timestamp                   9424 non-null   object 
 4   source                      9424 non-null   object 
 5   text                        9424 non-null   object 
 6   retweeted_status_id         724 non-null    float64
 7   retweeted_status_user_id    724 non-null    float64
 8   retweeted_status_timestamp  724 non-null    object 
 9   expanded_urls               9188 non-null   object 
 10  rating_numerator            9424 non-null   int64  
 11  rating_denominator          9424 non-null   int64  
 12  name                        9424 non-null   object 
 13  favorite_count              9300 

In [430]:
twitter_archive_clean.tweet_id.duplicated().sum()

7068

In [431]:
twitter_archive_clean.duplicated().sum()

0

### Quality Issues

#### Kindly note that some identical quality issues were merged and cleaned together

    1.'in_reply_to_status','in_reply_to_user_id','retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp' all have missing data.

Drop these columns as they do not provide useful insights.

In [432]:
twitter_archive_clean.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator',
       'rating_denominator', 'name', 'favorite_count', 'retweet_count',
       'dog_stage', 'value'],
      dtype='object')

In [433]:
columns_drop = ['in_reply_to_status_id', 'in_reply_to_user_id','text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp']
twitter_archive_clean.drop(columns_drop, axis= 1, inplace= True)

In [434]:
twitter_archive_clean.columns

Index(['tweet_id', 'timestamp', 'source', 'expanded_urls', 'rating_numerator',
       'rating_denominator', 'name', 'favorite_count', 'retweet_count',
       'dog_stage', 'value'],
      dtype='object')

    2. Missing data in 'expanded_urls' column.
drop column


In [435]:
twitter_archive_clean.drop('expanded_urls', axis= 1, inplace= True)

In [436]:
twitter_archive_clean.columns

Index(['tweet_id', 'timestamp', 'source', 'rating_numerator',
       'rating_denominator', 'name', 'favorite_count', 'retweet_count',
       'dog_stage', 'value'],
      dtype='object')

    3.Data type for 'timestamp' column should be date_time.
 change data type.

In [437]:
twitter_archive_clean.timestamp = pd.to_datetime(twitter_archive_clean['timestamp'])

In [438]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9424 entries, 0 to 9423
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            9424 non-null   int64              
 1   timestamp           9424 non-null   datetime64[ns, UTC]
 2   source              9424 non-null   object             
 3   rating_numerator    9424 non-null   int64              
 4   rating_denominator  9424 non-null   int64              
 5   name                9424 non-null   object             
 6   favorite_count      9300 non-null   float64            
 7   retweet_count       9300 non-null   float64            
 8   dog_stage           9424 non-null   object             
 9   value               9424 non-null   object             
dtypes: datetime64[ns, UTC](1), float64(2), int64(3), object(4)
memory usage: 736.4+ KB


    4. 'source' column data are mostly duplicated.
 drop source column

In [439]:
twitter_archive_clean.drop('source', axis= 1, inplace= True)

In [440]:
twitter_archive_clean.columns

Index(['tweet_id', 'timestamp', 'rating_numerator', 'rating_denominator',
       'name', 'favorite_count', 'retweet_count', 'dog_stage', 'value'],
      dtype='object')

    5.'Rating_denominator' have values greater other than 10.
Drop all the rows with denominator_rating other than 10.

In [441]:
twitter_archive_clean.query('rating_denominator != 10')

tweet_id                 timestamp  rating_numerator  \
313   835246439529840640 2017-02-24 21:54:03+00:00               960   
342   832088576586297345 2017-02-16 04:45:50+00:00                11   
433   820690176645140481 2017-01-15 17:52:40+00:00                84   
516   810984652412424192 2016-12-19 23:06:23+00:00                24   
784   775096608509886464 2016-09-11 22:20:06+00:00                 9   
...                  ...                       ...               ...   
8730  682962037429899265 2016-01-01 16:30:13+00:00                 7   
8731  682808988178739200 2016-01-01 06:22:03+00:00                20   
8847  677716515794329600 2015-12-18 05:06:23+00:00               144   
8911  675853064436391936 2015-12-13 01:41:41+00:00                88   
9403  666287406224695296 2015-11-16 16:11:11+00:00                 1   

      rating_denominator    name  favorite_count  retweet_count dog_stage  \
313                    0    None          1991.0           63.0     doggo   
342                   15    None            59.0            2.0     doggo   
433                   70    None         11579.0         2997.0     doggo   
516                    7     Sam          5055.0         1330.0     doggo   
784                   11    None             NaN            NaN     doggo   
...                  ...     ...             ...            ...       ...   
8730                  11  Darrel         33797.0        14889.0     puppo   
8731                  16    None          1629.0          154.0     puppo   
8847                 120    None          2784.0          877.0     puppo   
8911                  80    None          2427.0         1135.0     puppo   
9403                   2      an           123.0           57.0     puppo   

     value  
313   None  
342   None  
433   None  
516   None  
784   None  
...    ...  
8730  None  
8731  None  
8847  None  
8911  None  
9403  None  

[92 rows x 9 columns]

In [442]:
wrong_denominator = list(twitter_archive_clean.query('rating_denominator != 10').index)
twitter_archive_clean.drop(index= wrong_denominator, inplace= True)

In [443]:
twitter_archive_clean.query('rating_denominator != 10')

Empty DataFrame
Columns: [tweet_id, timestamp, rating_numerator, rating_denominator, name, favorite_count, retweet_count, dog_stage, value]
Index: []

    6.Very high rating 'rating_numerator of 1776 @ row 979.
drop this row.

In [444]:
twitter_archive_clean.query('rating_numerator == 1776')

tweet_id                 timestamp  rating_numerator  \
979   749981277374128128 2016-07-04 15:00:45+00:00              1776   
3335  749981277374128128 2016-07-04 15:00:45+00:00              1776   
5691  749981277374128128 2016-07-04 15:00:45+00:00              1776   
8047  749981277374128128 2016-07-04 15:00:45+00:00              1776   

      rating_denominator     name  favorite_count  retweet_count dog_stage  \
979                   10  Atticus          4784.0         2278.0     doggo   
3335                  10  Atticus          4784.0         2278.0   floofer   
5691                  10  Atticus          4784.0         2278.0    pupper   
8047                  10  Atticus          4784.0         2278.0     puppo   

     value  
979   None  
3335  None  
5691  None  
8047  None

In [445]:
ind = [979, 3335, 5691, 8047]
twitter_archive_clean.drop(index= ind, inplace= True)

In [446]:
twitter_archive_clean.query('rating_numerator == 1776')

Empty DataFrame
Columns: [tweet_id, timestamp, rating_numerator, rating_denominator, name, favorite_count, retweet_count, dog_stage, value]
Index: []

    7. Dog_name consists of missing data and other names such as 'a' and 'an'.
   replace the missing dog names with 'None' using pandas replace function.

In [447]:
twitter_archive_clean.name.value_counts()

None          2912
a              216
Charlie         48
Lucy            44
Cooper          44
              ... 
Edmund           4
Aqua             4
Chase            4
Meatball         4
Christoper       4
Name: name, Length: 954, dtype: int64

In [448]:
twitter_archive_clean['name'] = twitter_archive_clean.name.replace(['a'], 'None')

In [449]:
twitter_archive_clean.name.value_counts()

None          3128
Charlie         48
Lucy            44
Oliver          44
Cooper          44
              ... 
Edmund           4
Aqua             4
Chase            4
Meatball         4
Christoper       4
Name: name, Length: 953, dtype: int64

In [450]:
twitter_archive_clean.query('name == "an"')
twitter_archive_clean['name'] = twitter_archive_clean.name.replace(['an'], 'None')

In [451]:
twitter_archive_clean.query('name == "an"')

Empty DataFrame
Columns: [tweet_id, timestamp, rating_numerator, rating_denominator, name, favorite_count, retweet_count, dog_stage, value]
Index: []

In [452]:
twitter_archive_clean.sample(5)

tweet_id                 timestamp  rating_numerator  \
1593  686386521809772549 2016-01-11 03:17:53+00:00                11   
426   821407182352777218 2017-01-17 17:21:47+00:00                14   
8944  675111688094527488 2015-12-11 00:35:44+00:00                 8   
3387  745422732645535745 2016-06-22 01:06:43+00:00                 9   
7235  859851578198683649 2017-05-03 19:26:06+00:00                13   

      rating_denominator      name  favorite_count  retweet_count dog_stage  \
1593                  10   Crimson          3009.0          791.0     doggo   
426                   10  Sundance         11097.0         3990.0     doggo   
8944                  10    Oliver           867.0          213.0     puppo   
3387                  10      very          7948.0         2202.0   floofer   
7235                  10   Trooper         13966.0         3067.0     puppo   

      value  
1593   None  
426   doggo  
8944   None  
3387   None  
7235   None

    8. Change data type for retweet and favorite count after merging from float to int.
remove the null values then change the data types

In [385]:
twitter_archive_clean

tweet_id                 timestamp  rating_numerator  \
0     892420643555336193 2017-08-01 16:23:56+00:00                13   
1     892177421306343426 2017-08-01 00:17:27+00:00                13   
2     891815181378084864 2017-07-31 00:18:03+00:00                12   
3     891689557279858688 2017-07-30 15:58:51+00:00                13   
4     891327558926688256 2017-07-29 16:00:24+00:00                12   
...                  ...                       ...               ...   
9419  666049248165822465 2015-11-16 00:24:50+00:00                 5   
9420  666044226329800704 2015-11-16 00:04:52+00:00                 6   
9421  666033412701032449 2015-11-15 23:21:54+00:00                 9   
9422  666029285002620928 2015-11-15 23:05:30+00:00                 7   
9423  666020888022790149 2015-11-15 22:32:08+00:00                 8   

      rating_denominator      name  favorite_count  retweet_count dog_stage  \
0                     10   Phineas         33780.0         6996.0     doggo   
1                     10     Tilly         29306.0         5295.0     doggo   
2                     10    Archie         22034.0         3475.0     doggo   
3                     10     Darla         36898.0         7217.0     doggo   
4                     10  Franklin         35280.0         7748.0     doggo   
...                  ...       ...             ...            ...       ...   
9419                  10      None            89.0           36.0     puppo   
9420                  10      None           247.0          115.0     puppo   
9421                  10      None           100.0           36.0     puppo   
9422                  10      None           112.0           39.0     puppo   
9423                  10      None          2290.0          422.0     puppo   

     value  
0     None  
1     None  
2     None  
3     None  
4     None  
...    ...  
9419  None  
9420  None  
9421  None  
9422  None  
9423  None  

[9328 rows x 9 columns]

In [386]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9328 entries, 0 to 9423
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            9328 non-null   int64              
 1   timestamp           9328 non-null   datetime64[ns, UTC]
 2   rating_numerator    9328 non-null   int64              
 3   rating_denominator  9328 non-null   int64              
 4   name                9328 non-null   object             
 5   favorite_count      9208 non-null   float64            
 6   retweet_count       9208 non-null   float64            
 7   dog_stage           9328 non-null   object             
 8   value               9328 non-null   object             
dtypes: datetime64[ns, UTC](1), float64(2), int64(3), object(3)
memory usage: 728.8+ KB


In [418]:
twitter_archive_clean.sample(5)

tweet_id                 timestamp  rating_numerator  \
8813  679148763231985668 2015-12-22 03:57:37+00:00                 8   
7668  798673117451325440 2016-11-15 23:44:44+00:00                13   
2920  802323869084381190 2016-11-26 01:31:31+00:00                13   
2928  801285448605831168 2016-11-23 04:45:12+00:00                10   
6368  683357973142474752 2016-01-02 18:43:31+00:00                10   

      rating_denominator     name  favorite_count  retweet_count dog_stage  \
8813                  10     None               0              0     puppo   
7668                  10     None               0              0     puppo   
2920                  10  Severus               0              0   floofer   
2928                  10     None               0              0   floofer   
6368                  10     None               0              0    pupper   

     value  
8813  None  
7668  None  
2920  None  
2928  None  
6368  None

In [391]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9328 entries, 0 to 9423
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            9328 non-null   int64              
 1   timestamp           9328 non-null   datetime64[ns, UTC]
 2   rating_numerator    9328 non-null   int64              
 3   rating_denominator  9328 non-null   int64              
 4   name                9328 non-null   object             
 5   favorite_count      9328 non-null   int32              
 6   retweet_count       9328 non-null   int32              
 7   dog_stage           9328 non-null   object             
 8   value               9328 non-null   object             
dtypes: datetime64[ns, UTC](1), int32(2), int64(3), object(3)
memory usage: 655.9+ KB


In [282]:
image_prediction_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [283]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9328 entries, 0 to 9423
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            9328 non-null   int64              
 1   timestamp           9328 non-null   datetime64[ns, UTC]
 2   rating_numerator    9328 non-null   int64              
 3   rating_denominator  9328 non-null   int64              
 4   name                9328 non-null   object             
 5   favorite_count      9328 non-null   int32              
 6   retweet_count       9328 non-null   int32              
 7   dog_stage           9328 non-null   object             
 8   value               9328 non-null   object             
dtypes: datetime64[ns, UTC](1), int32(2), int64(3), object(3)
memory usage: 913.9+ KB


In [303]:
twitter_archive_clean.sample(10)

tweet_id                 timestamp  rating_numerator  \
2914  803321560782307329 2016-11-28 19:35:59+00:00                11   
296   837366284874571778 2017-03-02 18:17:34+00:00                13   
4526  669351434509529089 2015-11-25 03:06:32+00:00                10   
977   750011400160841729 2016-07-04 17:00:26+00:00                11   
1200  716730379797970944 2016-04-03 20:53:33+00:00                12   
40    884562892145688576 2017-07-11 00:00:02+00:00                13   
6358  683834909291606017 2016-01-04 02:18:42+00:00                 9   
5128  822610361945911296 2017-01-21 01:02:48+00:00                12   
5095  827199976799354881 2017-02-02 17:00:17+00:00                13   
726   782598640137187329 2016-10-02 15:10:30+00:00                11   

      rating_denominator      name  favorite_count  retweet_count dog_stage  \
2914                  10  Philbert               0              0   floofer   
296                   10      Lucy               1              1     doggo   
4526                  10     Sully               0              0   floofer   
977                   10     Piper               0              0     doggo   
1200                  10      None               0              0     doggo   
40                    10     Kevin               0              0     doggo   
6358                  10      None               0              0    pupper   
5128                  10      None               0              0    pupper   
5095                  10   Charlie               0              0    pupper   
726                   10     Timmy               0              0     doggo   

       value  
2914    None  
296     None  
4526    None  
977    doggo  
1200    None  
40      None  
6358  pupper  
5128    None  
5095    None  
726     None